test/persistent_receiver_test.go

```diff
@@ -1696,19 +1694,45 @@ var _ = Describe("PersistentReceiver", func() {
 
 				It("handles unsolicited terminations while already terminating", func() {
 					blocker := make(chan struct{})
+					signal := make(chan bool)
 					messageChannel := make(chan message.InboundMessage, numMessages+1)
+					waitToReceiveMessages := 30 * time.Second // This timer needs to be big enough to give us a very
+					// high chance of both receiving the messages and getting
+					// past the state change in Terminate. It may need to be
+					// updated in the future since this only mitigates the
+					// race condition and doesn't eliminate it.
 					messageReceiver.ReceiveAsync(func(inboundMessage message.InboundMessage) {
+					loop:
+						for timeout := time.After(waitToReceiveMessages); ; {
+							select {
+							case <-timeout:
+								break loop
+							default:
+							}
+							if !messageReceiver.IsRunning() {
+								break loop
+							}
+						}
+						signal <- true
 						<-blocker
 						messageChannel <- inboundMessage
 					})
 					helpers.PublishNPersistentMessages(messagingService, topicString, numMessages+1)
 					helpers.ValidateMetric(messagingService, metrics.PersistentMessagesReceived, numMessages+1)
-					terminateDuration := 2 * time.Second
+					terminateDuration := 2 * time.Second // This timer does not need to be especially long, since we're
+					// only needed a fraction of a second to block on signal
+					// and wait for that channel to be written to in the receiver
+					// callback.
 					terminateChan := messageReceiver.TerminateAsync(terminateDuration)
+					// Blocking on this signal prevents us from proceeding with the test unless either the timer has
+					// expired or the state transition has occured.
+					<-signal
+					Expect(messageReceiver.IsRunning()).To(BeFalse())
```

Comment:
@TrentDaniel I try to focus on the order of events to test here:
1. create receiver and Publish N messages to it.
2. Receiver first message and block
3. Start termination given terminateChan
4. Ensure termination is in progress, (verify not running behaviour)
5. terminate receiver independent of 3. (in this case remove queue via semp? or Otherwise) (verify behaviour not running behaviour)
6. Ensure termination duration for 3. has expired.
7. unblock receive callback.
8. Ensure terminate from 3. completes and receiver callback has finished.